docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,838 @@
+  has been initialized by the callee.
+  Note that the caller should not initialize the content of `nb` before invoking `foo`. Any content
+  set before the invocation will not be passed to the callee (i.e., by the definition of an `out`
+  parameter) and the content will be overwritten after a successful invocation.
+
+  ```
+  NestedBlob nb;
+  memset(&nb, 0, sizeof(NestedBlob));
+  // b->len and b->buf have not been initialized yet.
+  foo(enclave, &nb); // The casae of an ECALL,
+                     // or simply foo(&nb) for the case of an OCALL.
+  ...
+  /* Caller is responsible for freeing the buffers after use. */
+  if (nb.blob_array)
+  {
+      for (size_t i = 0; i < nb.num; i++)
+          free(nb.blob_array[i].buf);
+      free(nb.blob_array);
+  }
+  ```
+
+  - Memory management
+  
+    The caller is responsible for managing the content of `nb` after invoking `foo`.
+    Each nested buffer of `nb` is allocated separately and is therefore required to
+    free each of the buffers as shown in the example.
+
+- Callee
+
+  The callee is responsible for initializing the `nb` if the `nb` is not `NULL`.
+  The `nb` being `NULL` indicates that the caller passes `NULL` to `foo`.
+  
+  ```
+  // User-defined code
+  void foo(NestedBlob *nb)
+  {
+      if (!nb)
+          return;
+      nb->num = 5;
+      nb->blob_array = (Blob*)malloc(5 * sizeof(Blob));
+      for (size_t i = 0; i < 5; i++)
+      {
+          Blob* b = &ptr->blob_array[i];
+          b->len = 10;
+          b->buf = (char*)malloc(10);
+          memset(b->buf, 'A', 10);
+      }
+  }
+  ```
+
+  - Memory management
+
+    The callee should **allocate each nested buffer separately on the heap** as shown in the example.
+    These allocated buffers are freed automatically by the oeedger8r-generated code and therefore the
+    callee does not need to take further actions. 
+    
+Note that the proposal does not affect the behavior of the existing deep-copy `inout` parameter.
+
+### Design
+
+The design assumes that the shared memory between the host and the enclave is available.
+The idea of the design is that after the callee (either a host or an enclave) initializes `nb`,
+the callee serializes the content of `nb` into a local buffer. Next, the content of the buffer
+is copied over to a caller's buffer. Given that only the enclave has access to the memory
+on the host, the enclave, either being the caller or callee, is always responsible for copying the buffer.
+Finally, the caller unserializes the received content, which fills up the `nb`.
+The rest of the section walks through modified oeedger8r-generated code for both ECALL and OCALL that realizes the design.
+
+Note that this design is compatible with switchelss call feature, which also requires shared memory.
+
+- ECALL
+
+  The design adds a `_host_buffer` member to the internal struct that holds the shared pointer.
```

Comment:
You could also add a `_host_buffer_size` member. The first 3 fields of any marshalling struct would then be:
```c
typedef struct _fcn_args_t
  {
      oe_result_t _result;
      uint8_t* _host_buffer;
      size_t _host_buffer_size;
      ... // other marshaling fields
  }
```

Reply:
Updated.